File: backend/app/models/requests/schema.py
from typing import Literal
from uuid import UUID
from pydantic import BaseModel, Field


ColumnType = Literal["string", "integer", "float", "date", "boolean"]


class DestinationColumnRequest(BaseModel):
    name: str = Field(min_length=1)
    display_name: str | None = None
    type: ColumnType
    nullable: bool
    date_format: str | None = None


class CreateSchemaRequest(BaseModel):
    project_id: UUID
    name: str = Field(min_length=1, max_length=100)
    columns: list[DestinationColumnRequest] = Field(min_length=1)


class UpdateSchemaRequest(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=100)
    columns: list[DestinationColumnRequest] | None = None

File: frontend/types/schema.ts
import { z } from "zod";

export const columnTypeSchema = z.enum(["string", "integer", "float", "date", "boolean"]);

export const destinationColumnSchema = z.object({
  name: z.string().min(1),
  display_name: z.string().optional(),
  type: columnTypeSchema,
  nullable: z.boolean(),
  date_format: z.string().optional(),
});

export const createSchemaSchema = z.object({
  name: z.string().min(1, "Name is required").max(100),
  columns: z.array(destinationColumnSchema).min(1, "At least one column is required"),
});

export type DestinationColumn = z.infer<typeof destinationColumnSchema>;
export type CreateSchemaInput = z.infer<typeof createSchemaSchema>;

export type Schema = {
  id: string;
  project_id: string;
  user_id: string;
  name: string;
  definition: {
    version: number;
    columns: DestinationColumn[];
  };
  created_at: string;
  updated_at: string;
};
